visualization/app/codeCharta/ui/legendPanel/legendMarkedPackages/legendMarkedPackages.component.tsx

```diff
@@ -0,0 +1,63 @@
+import "./legendMarkedPackages.component.scss"
+import { IRootScopeService } from "angular"
+import { CodeMapActionsService } from "../../codeMap/codeMap.actions.service"
+import { MarkedPackage } from "../../../codeCharta.model"
+import { MarkedPackagesSubscriber, MarkedPackagesService } from "../../../state/store/fileSettings/markedPackages/markedPackages.service"
+import { hasValidHexLength, normalizeHex } from "../../customColorPicker/colorHelper"
+
+type Path = string
+interface MarkedPackageMap {
+	[color: string]: Path[]
+}
+
+export class LegendPanelController implements MarkedPackagesSubscriber {
+	private _viewModel: {
+		packageLists: MarkedPackageMap
+	} = {
+		packageLists: {}
+	}
+
+	private _colorPickerScopes = {}
+
+	constructor(private $rootScope: IRootScopeService, private $scope, private codeMapActionsService: CodeMapActionsService) {
+		this.$scope._colorPickerScopes = {}
+		MarkedPackagesService.subscribe(this.$rootScope, this)
+	}
+
+	onMarkedPackagesChanged(markedPackages: MarkedPackage[]) {
+		console.log(this.$scope, this._colorPickerScopes)
+		this._viewModel.packageLists = markedPackages.reduce((acc, markedPackage) => {
+			if (!acc.hasOwnProperty(markedPackage.color)) acc[markedPackage.color] = []
+			acc[markedPackage.color].push(markedPackage.path)
+			return acc
+		}, {})
+
+		this._colorPickerScopes = Object.entries(this._viewModel.packageLists).reduce((acc, [color, paths]) => {
+			acc[paths[0]] = color
+			return acc
+		}, {})
+	}
+
+	$onInit() {
+		this.$scope.colorPickerEventApi = {
+			onChange: (api, newColor) => {
+				if (!hasValidHexLength(newColor)) return
+
+				newColor = normalizeHex(newColor)
+				const oldColor = api.getElement().scope().color
+				this._viewModel.packageLists[newColor] = this._viewModel.packageLists[oldColor]
+				delete this._viewModel.packageLists[oldColor]
+
+				this._viewModel.packageLists[newColor].forEach(path => {
+					this.codeMapActionsService.markFolder({ path }, newColor)
+				})
```

Comment:
I suggest to use a map, if possible. Keep it, if it is more than a few minutes of work to change it :)

Reply:
Fascinating: In the context of JS programming, it never occurred to me to use a `Map`. But of course, that would be nicer to use. Now that it occurred to me once, I will remember it for the next time :)